text/0000-argument-lifetimes.md

```diff
@@ -0,0 +1,476 @@
+
+**Note: these changes are designed to *not* require a new epoch**. They
+introduce several deprecations, which are essentially style lints. The next
+epoch should turn these deprecations into errors.
+
+## Lifetimes in `impl` headers
+
+When writing an `impl` header, it is deprecated to bind a lifetime parameter
+within the generics specification (e.g. `impl<'a>`). Instead the `impl` header
+can mention lifetimes without adding them as generics. Any lifetimes that are
+not already in scope (which, today, means any lifetime whatsoever) is treated as
+being bound as a parameter of the `impl`.
+
+Thus, where today you would write:
+
+```rust
+impl<'a> Iterator for MyIter<'a> { ... }
+impl<'a, 'b> SomeTrait<'a> for SomeType<'a, 'b> { ... }
+```
+
+tomorrow you would write:
+
+```rust
+impl Iterator for MyIter<'a> { ... }
+impl SomeTrait<'a> for SomeType<'a, 'b> { ... }
+```
+
+## Lifetimes in `fn` signatures
+
+When writing a `fn` declaration, it is deprecated to bind a lifetime parameter
+within the generics specification (e.g. `fn foo<'a>(arg: &'a str)`).
+
+Instead:
+
+- If a lifetime appears that is not already in scope, it is taken to be a new
+  binding, treated as a parameter to the function.
+- If an argument has exactly one elided lifetime, you can refer to that lifetime
+  by the argument's name preceded by `'`. Otherwise, it is not permitted to use
+  that lifetime name (unless it is already in scope, which generates a warning).
+- As with today's elision rules, lifetimes that appear *only* within `Fn`-style
+  bounds or trait object types are bound in higher-rank form (i.e., as if you'd
+  written them using a `for<'a>`).
+
+Thus, where today you would write:
+
+```rust
+fn elided(&self) -> &str;
+fn two_args<'b>(arg1: &Foo, arg2: &'b Bar) -> &'b Baz
+fn two_lifetimes<'a, 'b: 'a>(arg1: &'a Foo, arg2: &'b Bar) -> &'a Quux<'b>
+
+impl<'a> MyStruct<'a> {
+    fn foo(&self) -> &'a str;
+    fn bar<'b>(&self, arg: &'b str) -> &'b str;
+}
+
+fn take_fn<'a>(x: &'a u32, y: for<'b> fn(&'a u32, &'b u32, &'b u32))
+```
+
+tomorrow you would write:
+
+```rust
+fn elided(&self) -> &str;
+fn two_args(arg1: &Foo, arg2: &Bar) -> &'arg2 Baz
+fn two_lifetimes(arg1: &Foo, arg2: &Bar) -> &'arg1 Quux<'arg2>
+
+impl MyStruct<'A> {
+    fn foo(&self) -> &'A str;
+    fn bar(&self, arg: &'b str) -> &'b str;
+}
+
+fn take_fn(x: &u32, y: fn(&'x u32, &'b u32, &'b u32));
+```
+
+## The wildcard lifetime
+
+When referring to a type (other than `&`/`&mut`) that requires lifetime
+arguments, it is deprecated to leave off those parameters.
+
+Instead, you can write a `'_` for the parameters, rather than giving a lifetime
+name, which will have identical behavior to leaving them off today.
+
+Thus, where today you would write:
+
+```rust
+fn foo(&self) -> Ref<SomeType>
+fn iter(&self) -> Iter<T>
+```
+
+tomorrow you would write:
+
+```rust
+fn foo(&self) -> Ref<'_, SomeType>
+fn iter(&self) -> Iter<'_, T>
+```
```

Comment:
It is not intended to be allowed. One would have to write `Foo<'_, '_>`. The intention is that, for a struct, you will always provide the full set of parameters, matching the struct declaration exactly, but using `'_` if you like to avoid giving an explicit name when one is not needed.